File: features/step_definitions/requests.rb
When /^I request \/(.*)$/ do |path|
  reprocess
  OSRMBackgroundLauncher.new("#{@osm_file}.osrm") do
    @response = request_path path
  end
end

Then /^I should get a response/ do
  @response.code.should == "200"
  @response.body.should_not == nil
  @response.body.should_not == ''
end

Then /^response should be valid JSON$/ do
  @json = JSON.parse @response.body
end

Then /^response should be well-formed$/ do
  @json['status'].class.should == Fixnum
end

Then /^status code should be (\d+)$/ do |code|
  @json = JSON.parse @response.body
  @json['status'].should == code.to_i
end

Then /^status message should be "(.*?)"$/ do |message|
  @json = JSON.parse @response.body
  @json['status_message'].should == message
end

Then /^response should be a well-formed route$/ do
  step "response should be well-formed"
  @json['status_message'].class.should == String
  @json['route_summary'].class.should == Hash
  @json['route_geometry'].class.should == String
  @json['route_instructions'].class.should == Array
  @json['via_points'].class.should == Array
  @json['via_indices'].class.should == Array
end

Then /^"([^"]*)" should return code (\d+)$/ do |binary, code|
  @process_error.is_a?(OSRMError).should == true
  @process_error.process.should == binary
  @process_error.code.to_i.should == code.to_i
end

File: features/step_definitions/requests.js
// Step definitions for generic HTTP request testing and response validation
import assert from 'assert';
import { When, Then } from '@cucumber/cucumber';

When(/^I request \/(.*)$/, async function (path) {
  await this.reprocessAndLoadData();
  await new Promise((resolve, reject) => {
    this.requestUrl(path, (err, res, body) => {
      if (err) return reject(err);
      this.response = res;
      this.body = body;
      resolve();
    });
  });
});

Then(/^I should get a response/, function () {
  this.ShouldGetAResponse();
});

Then(/^response should be valid JSON$/, function (callback) {
  this.ShouldBeValidJSON(callback);
});

Then(/^response should be well-formed$/, function () {
  this.ShouldBeWellFormed();
});

Then(/^status code should be (.+)$/, function (code, callback) {
  try {
    this.json = JSON.parse(this.body);
  } catch(e) {
    return callback(e);
  }
  assert.equal(this.json.code, code);
  callback();
});

Then(/^status message should be "(.*?)"$/, function (message, callback) {
  try {
    this.json = JSON.parse(this.body);
  } catch(e) {
    return callback(e);
  }
  assert(this.json.status_message, message);
  callback();
});

Then(/^response should be a well-formed route$/, function () {
  this.ShouldBeWellFormed();
  assert.equal(this.json.code, 'ok');
  assert.ok(Array.isArray(this.json.routes));
  assert.ok(Array.isArray(this.json.waypoints));
});

Then(/^"([^"]*)" should return code (\d+)$/, function (binary, code) {
  assert.ok(this.processError instanceof Error);
  assert.equal(this.processError.process, binary);
  assert.equal(parseInt(this.processError.code), parseInt(code));
});